### PR TITLE
Raise exceptions in `Server.handle_stream` instead of swallowing/logging.

### PR DESCRIPTION
**Pull request by hendrikmakait**
When closing the stream handler, it would be helpful to understand the reason instead of stopping silently. 

In particular, we should not swallow an `OSError` silently when it may lead to the comm being closed.

- [x] Tests added / passed
- [x] Passes `pre-commit run --all-files`

### CODE REVIEW

File: distributed/core.py [hunk 1]
```diff
@@ -825,12 +825,16 @@ async def handle_stream(self, comm, extra=None):
...
+                        if msg == "OK":
                             break
                         op = msg.pop("op")
                         if op:
                             if op == "close-stream":
                                 closed = True
+                                logger.info(
+                                    "Received 'close-stream' from %s; closing.",
+                                    comm.peer_address,
+                                )
```
Comment (fjetter) [inline distributed/core.py:837]: This is a bit verbose, isn't it?

File: distributed/core.py [hunk 2]
```diff
@@ -843,16 +847,23 @@ async def handle_stream(self, comm, extra=None):
                         else:
                             logger.error("odd message %s", msg)
                     await asyncio.sleep(0)
-
+        except CommClosedError:
+            logger.info("Connection to %s already closed.", comm.peer_address)
```
Comment (fjetter) [inline distributed/core.py:851]: Is this actually worth a log?

**Comment by hendrikmakait** [comment]
After an offline discussion with @fjetter, we decided that `handle_stream` should raise all exceptions and leave handling them to the caller.

File: distributed/core.py (hunk 1 above)
Comment (hendrikmakait) [reply distributed/core.py:837]: `close-stream` should only occur in a few instances, I would like to be able to see when those happen.

File: distributed/worker.py [hunk 3]
```diff
@@ -1265,14 +1266,14 @@ async def heartbeat(self) -> None:
...
-            "Connection to scheduler broken. Closing without reporting. ID: %s Address %s Status: %s",
-            self.id,
-            self.address,
-            self.status,
-        )
-        await self.close(reason="worker-handle-scheduler-connection-broken")
+        try:
+            await self.handle_stream(comm)
+        except CommClosedError:
+            logger.info(
```
Comment (fjetter) [inline distributed/worker.py:1272]: 1. Should be error, shouldn't it?
2. Why not reraise?

File: distributed/scheduler.py [hunk 4]
```diff
@@ -5415,6 +5415,8 @@ async def handle_worker(self, comm: Comm, worker: str) -> None:
         logger.info("Starting worker compute stream, %s", worker)
         try:
             await self.handle_stream(comm=comm, extra={"worker": worker})
+        except CommClosedError:
+            logger.info("Connection to worker %s has been closed.", comm.peer_address)
```
Comment (fjetter) [inline distributed/scheduler.py:5419]: exception level + reraise?

File: distributed/worker.py (hunk 3 above)
Comment (hendrikmakait) [reply distributed/worker.py:1264]: I was under the impression that this is part of "normal" shutdown. After revisiting the code, it looks like this should not happen and would rather be a concurrency issue that should be properly dealt with.

File: distributed/scheduler.py (hunk 4 above)
Comment (fjetter) [reply distributed/scheduler.py:5419]: Why not reraise here?

Comment (hendrikmakait) [reply distributed/scheduler.py]: My bad, apparently I forgot to push that change. It's there now.

File: distributed/core.py [hunk 5]
```diff
@@ -815,40 +815,48 @@ async def _handle_comm(self, comm):
...
+        logger.info("Starting established connection to %s", comm.peer_address)
 
         closed = False
         try:
             while not closed:
-                msgs = await comm.read()
+                try:
+                    msgs = await comm.read()
+                # If another coroutine has closed the comm, stop handling the stream.
+                except CommClosedError:
```
Comment (hendrikmakait) [inline distributed/core.py:826]: As discussed offline with @fjetter, we should use the `CommClosedError` as a stopping and synchronization mechanism between coroutines that are dealing with the comm. If one of them decides to close the comm, the `CommClosedError` allows us to propagate that without race conditions.

File: distributed/scheduler.py (hunk 4 above)
Comment (hendrikmakait) [reply distributed/scheduler.py]: Superseded by https://github.com/dask/distributed/pull/7162#discussion_r1011638692

File: distributed/worker.py (hunk 3 above)
Comment (hendrikmakait) [reply distributed/worker.py]: Superseded by https://github.com/dask/distributed/pull/7162#discussion_r1011638692

File: distributed/core.py (hunk 5 above)
Comment (fjetter) [reply distributed/core.py:826]: FWIW It's not ideal but I believe this is how it was intended at some point. We may need to revise this but that would be a bigger change that is feasible right now.

**Review by fjetter — APPROVED** [review]

**Pull request merged**